src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs

```diff
@@ -296,5 +284,140 @@ internal SafeSslHandle(IntPtr validSslPointer, bool ownsHandle) : base(IntPtr.Ze
         {
             handle = validSslPointer;
         }
+
+        internal class ReadBioBuffer : IDisposable
+        {
+            private SafeBioHandle _bioHandle;
+            private GCHandle _handle;
+            private int _bytesAvailable;
+            private byte[] _byteArray;
+            private int _offset;
+
+            internal ReadBioBuffer(SafeBioHandle bioHandle)
+            {
+                _bioHandle = bioHandle;
+                _handle = GCHandle.Alloc(this, GCHandleType.Normal);
+                Interop.CustomBio.BioSetGCHandle(_bioHandle, _handle);
+                Interop.Crypto.BioSetShoudRetryReadFlag(bioHandle);
+            }
+
+            public void SetBio(byte[] buffer, int offset, int length)
+            {
+                _byteArray = buffer;
```

Comment:
Assert is a good idea, should never happen, the data is staged by the higher up SslStreamInternal, it reads a header and ensures that a frame has been read into that buffer.